src/fixed.c

```diff
@@ -9,42 +9,70 @@
 #define uint32_lo(i) ((i) & 0xffff)
 #define uint32_hi(i) ((i) >> 16)
 #define uint32_carry16 ((1) << 16)
+/* Check interval
+ * For any variable interval checking:
+ *     if (x > minx - epsilon && x < minx + epsilon) ...
+ * it is faster to use bit operation:
+ *     if ((signed)((x - (minx - epsilon)) | ((minx + epsilon) - x)) > 0) ...
+ */
+#define CHECK_INTERVAL(x, minx, epsilon) \
+    ((int32_t) ((x - (minx - epsilon)) | (minx + epsilon - x)) > 0)
 
 twin_fixed_t twin_fixed_sqrt(twin_fixed_t a)
 {
-    twin_fixed_t max = a, min = 0;
-
-    while (max > min) {
-        twin_fixed_t mid = (max + min) >> 1;
-        if (mid >= 181 * TWIN_FIXED_ONE) {
-            max = mid - 1;
-            continue;
-        }
-        twin_fixed_t sqr = twin_fixed_mul(mid, mid);
-        if (sqr == a)
-            return mid;
-        if (sqr < a)
-            min = mid + 1;
-        else
-            max = mid - 1;
+    if (a <= 0)
+        return 0;
+    if (CHECK_INTERVAL(a, TWIN_FIXED_ONE, (1 << (8 - 1))))
+        return TWIN_FIXED_ONE;
+
+    /* count leading zero */
+    int offset = 0;
+    for (twin_fixed_t i = a; !(0x40000000 & i); i <<= 1) {
+        ++offset;
     }
-    return (max + min) >> 1;
+    /* shift left 'a' to expand more digit for sqrt precision */
+    offset &= ~1;
+    a <<= offset;
+    /* calculate the digits need to shift back */
+    offset >>= 1;
+    offset -= (16 >> 1);
+    /* Use digit-by-digit calculation to compute square root */
+    twin_fixed_t z = 0;
+    for (twin_fixed_t m = 1UL << ((31 - __builtin_clz(a)) & ~1UL); m; m >>= 2) {
+        int b = z + m;
+        z >>= 1;
+        if (a >= b)
+            a -= b, z += m;
+    }
+    /* shift back the expanded digits */
+    return (offset >= 0) ? z >> offset : z << (-offset);
 }
 
 twin_sfixed_t _twin_sfixed_sqrt(twin_sfixed_t as)
 {
-    twin_dfixed_t max = as, min = 0;
-    twin_dfixed_t a = twin_sfixed_to_dfixed(as);
-
-    while (max > min) {
-        twin_dfixed_t mid = (max + min) >> 1;
-        twin_dfixed_t sqr = mid * mid;
-        if (sqr == a)
-            return (twin_sfixed_t) mid;
-        if (sqr < a)
-            min = mid + 1;
-        else
-            max = mid - 1;
+    if (as <= 0)
+        return 0;
+    if (CHECK_INTERVAL(as, TWIN_SFIXED_ONE, (1 << (2 - 1))))
+        return TWIN_SFIXED_ONE;
```

Comment:
I'm curious about the reason for using CHECK_INTERVAL.
![new_w](https://github.com/user-attachments/assets/adcd1bd4-7d1c-4e52-ba81-9451a487fbfb)
![new_wo](https://github.com/user-attachments/assets/2d298840-8be9-408a-bf9f-7d4d6fd195f9)

Reply:
This method is used to find an approximation of sin and cos values. However, this method makes it hard to calculate the square root when the numbers are near 1, which causes the wrong approximation of sin and cos values, especially when the radian is near pi/2 or 0.

Reply:
The two figures below are sin and cos value calculted with (left) and without (right) CHECK_INTERVAL. Test on [rv32emu/tests/line.c](https://github.com/sysprog21/rv32emu/blob/master/tests/line.c).
![line](https://github.com/user-attachments/assets/e6f24f2a-ff27-4f21-a9dd-fe4d561b589f)
![line_wrong](https://github.com/user-attachments/assets/66f8ef38-8213-4200-8269-208a8c27b74d)